src/workerd/api/global-scope.c++

```diff
@@ -347,21 +347,28 @@ void ServiceWorkerGlobalScope::sendTraces(kj::ArrayPtr<kj::Own<Trace>> traces,
     Worker::Lock& lock, kj::Maybe<ExportedHandler&> exportedHandler) {
   auto isolate = lock.getIsolate();
 
-  auto event = jsg::alloc<TraceEvent>(traces);
-
   KJ_IF_MAYBE(h, exportedHandler) {
-    KJ_IF_MAYBE(f, h->trace) {
-      auto promise = (*f)(lock, event->getTraces(), h->env.addRef(isolate), h->getCtx(isolate));
-      event->waitUntil(kj::mv(promise));
+    KJ_IF_MAYBE(f, h->tail) {
+      auto tailEvent = jsg::alloc<TailEvent>("tail"_kj, traces);
+      auto promise = (*f)(lock, tailEvent->getEvents(), h->env.addRef(isolate), h->getCtx(isolate));
+      tailEvent->waitUntil(kj::mv(promise));
+    } else KJ_IF_MAYBE(f, h->trace) {
+      auto traceEvent = jsg::alloc<TailEvent>("trace"_kj, traces);
+      auto promise = (*f)(lock, traceEvent->getEvents(), h->env.addRef(isolate), h->getCtx(isolate));
+      traceEvent->waitUntil(kj::mv(promise));
```

Comment:
I'm not sure, but we may want to log a warning to the user that this is a legacy op.

Reply:
Not really convinced it's worthwhile. The number of workers that will hit this path should be very minimal and if they continue to hit it they're unlikely be updated too often. The warning would just end up being noise that is unlikely to be acted upon.